File: plump/env/engine.py
# plump/env/engine.py — pure game engine, no torch, no encode.
# All house rules live here. No training logic.
from __future__ import annotations

import random
from .cards import SUIT_INDEX, SUITS


class PlumpEnv:
    """4-player trick-taking engine.

    House v2 rules:
    - 0-bid success = 5 pts (not 10) — scored outside, but noted here.
    - Highest bidder (earliest on tie) leads first trick.
    - No trump. Must follow suit. Void tracking.
    """

    SUITS = SUITS

    def __init__(self, num_players: int = 4):
        self.num_players = num_players
        self.hands: list[list[tuple[str, int]]] = []
        self.bids: list[int] = []
        self.tricks_won: list[int] = []
        self.table: list[tuple[int, tuple[str, int]]] = []
        self.led_suit: str = ""
        self.round_cards: int = 0
        self.current_trick: int = 0
        self.played_cards: set[tuple[str, int]] = set()
        # void_matrix[player][suit_idx] = True once proven void
        self.void_matrix: list[list[bool]] = [[False] * 4 for _ in range(self.num_players)]
        # history for tokenizer — ordered events (kind, player, payload)
        # kind: "bid" -> payload int, "play" -> payload card tuple
        self.history: list[tuple[str, int, object]] = []

    # -- setup --
    def new_round(self, round_cards: int = 10) -> None:
        self.round_cards = round_cards
        self.current_trick = 0
        self.table = []
        self.led_suit = ""
        self.played_cards = set()
        self.hands = [[] for _ in range(self.num_players)]
        self.bids = [-1] * self.num_players
        self.tricks_won = [0] * self.num_players
        self.void_matrix = [[False] * 4 for _ in range(self.num_players)]
        self.history = []

        deck = [(s, v) for s in self.SUITS for v in range(2, 15)]
        random.shuffle(deck)
        for i in range(round_cards * self.num_players):
            self.hands[i % self.num_players].append(deck[i])

    # -- queries --
    def can_follow_suit(self, player: int) -> bool:
        if self.led_suit == "":
            return False
        return any(c[0] == self.led_suit for c in self.hands[player])

    def is_legal(self, player: int, card: tuple[str, int]) -> bool:
        if card not in self.hands[player]:
            return False
        if self.led_suit == "" or card[0] == self.led_suit:
            return True
        return not self.can_follow_suit(player)

    def legal_cards(self, player: int) -> list[tuple[str, int]]:
        return [c for c in self.hands[player] if self.is_legal(player, c)]

    def legal_bids(self, player: int) -> list[int]:
        bids_made = sum(1 for b in self.bids if b != -1)
        is_last = bids_made == self.num_players - 1
        bid_sum = sum(b for b in self.bids if b != -1)
        forbidden = self.round_cards - bid_sum
        result: list[int] = []
        for b in range(0, self.round_cards + 1):
            if is_last and b == forbidden:
                continue
            result.append(b)
        return result

    def get_leader(self) -> int:
        """Earliest highest bidder leads. Fallback 0 if bids incomplete."""
        if any(b == -1 for b in self.bids):
            return 0
        max_bid = max(self.bids)
        for i, b in enumerate(self.bids):
            if b == max_bid:
                return i
        return 0

    # -- helpers to record history (decoupled from logic) --
    def record_bid(self, player: int, bid: int) -> None:
        self.history.append(("bid", player, bid))

    # -- mutations --
    def play_card(self, player: int, card: tuple[str, int]) -> None:
        if self.led_suit == "":
            self.led_suit = card[0]
        elif card[0] != self.led_suit:
            self.void_matrix[player][SUIT_INDEX[self.led_suit]] = True
        self.hands[player].remove(card)
        self.table.append((player, card))
        self.history.append(("play", player, card))

    def resolve_trick(self) -> tuple[int, list[tuple[int, tuple[str, int]]]]:
        if not self.table:
            raise ValueError("resolve_trick called with empty table")
        led_suit = self.table[0][1][0]
        winner = self.table[0][0]
        best_val = self.table[0][1][1]
        for p, c in self.table[1:]:
            if c[0] == led_suit and c[1] > best_val:
                best_val = c[1]
                winner = p
        self.tricks_won[winner] += 1
        result = (winner, list(self.table))
        for _, card in self.table:
            self.played_cards.add(card)
        self.table = []
        self.led_suit = ""
        self.current_trick += 1
        return result

File: crates/plump-engine/src/engine.rs
//! engine.rs — mirrors plump/env/engine.py, pure, no deps beyond rand.

use rand::seq::SliceRandom;
use rand::thread_rng;

use crate::cards::{card_index, SUITS};

#[derive(Clone, Debug)]
pub struct PlumpEnv {
    pub num_players: usize,
    pub hands: Vec<Vec<(char, u8)>>,
    pub bids: Vec<i8>,
    pub tricks_won: Vec<u8>,
    pub table: Vec<(usize, (char, u8))>,
    pub led_suit: Option<char>,
    pub round_cards: usize,
    pub current_trick: usize,
    pub played: Vec<(char, u8)>,
    pub void_matrix: Vec<[bool; 4]>,
    pub history: Vec<(String, usize, String)>, // stub for tokenizer
}

impl PlumpEnv {
    pub fn new() -> Self {
        Self {
            num_players: 4,
            hands: vec![],
            bids: vec![],
            tricks_won: vec![],
            table: vec![],
            led_suit: None,
            round_cards: 0,
            current_trick: 0,
            played: vec![],
            void_matrix: vec![[false; 4]; 4],
            history: vec![],
        }
    }

    pub fn new_round(&mut self, round_cards: usize) {
        self.round_cards = round_cards;
        self.current_trick = 0;
        self.table.clear();
        self.led_suit = None;
        self.played.clear();
        self.hands = vec![vec![]; 4];
        self.bids = vec![-1; 4];
        self.tricks_won = vec![0; 4];
        self.void_matrix = vec![[false; 4]; 4];
        self.history.clear();
        let mut deck: Vec<(char, u8)> = SUITS.iter().flat_map(|&s| (2..15).map(move |r| (s, r))).collect();
        deck.shuffle(&mut thread_rng());
        for i in 0..round_cards * 4 {
            self.hands[i % 4].push(deck[i]);
        }
    }

    pub fn can_follow(&self, player: usize) -> bool {
        if let Some(led) = self.led_suit {
            self.hands[player].iter().any(|(s, _)| *s == led)
        } else {
            false
        }
    }

    pub fn is_legal(&self, player: usize, card: (char, u8)) -> bool {
        if !self.hands[player].contains(&card) {
            return false;
        }
        match self.led_suit {
            None => true,
            Some(led) if card.0 == led => true,
            _ => !self.can_follow(player),
        }
    }

    pub fn legal_cards(&self, player: usize) -> Vec<(char, u8)> {
        self.hands[player].iter().copied().filter(|&c| self.is_legal(player, c)).collect()
    }

    pub fn legal_bids(&self, player: usize) -> Vec<i8> {
        let made = self.bids.iter().filter(|&&b| b != -1).count();
        let is_last = made == 3;
        let sum: i32 = self.bids.iter().filter(|&&b| b != -1).map(|&b| b as i32).sum();
        let forbidden = self.round_cards as i32 - sum;
        (0..=self.round_cards as i8)
            .filter(|&b| !(is_last && b as i32 == forbidden))
            .collect()
    }

    pub fn get_leader(&self) -> usize {
        if self.bids.iter().any(|&b| b == -1) {
            return 0;
        }
        let max = *self.bids.iter().max().unwrap();
        self.bids.iter().position(|&b| b == max).unwrap_or(0)
    }

    pub fn play_card(&mut self, player: usize, card: (char, u8)) {
        if self.led_suit.is_none() {
            self.led_suit = Some(card.0);
        } else if Some(card.0) != self.led_suit {
            let idx = match self.led_suit.unwrap() {
                'H' => 0,
                'S' => 1,
                'D' => 2,
                'C' => 3,
                _ => 0,
            };
            self.void_matrix[player][idx] = true;
        }
        self.hands[player].retain(|&c| c != card);
        self.table.push((player, card));
        self.history.push(("play".into(), player, format!("{}{}", card.0, card.1)));
    }

    pub fn resolve_trick(&mut self) -> (usize, Vec<(usize, (char, u8))>) {
        assert!(!self.table.is_empty());
        let led = self.table[0].1 .0;
        let mut winner = self.table[0].0;
        let mut best = self.table[0].1 .1;
        for &(p, (s, r)) in &self.table[1..] {
            if s == led && r > best {
                best = r;
                winner = p;
            }
        }
        self.tricks_won[winner] += 1;
        let res = (winner, self.table.clone());
        for &(_, c) in &self.table {
            self.played.push(c);
        }
        self.table.clear();
        self.led_suit = None;
        self.current_trick += 1;
        res
    }
}

impl Default for PlumpEnv {
    fn default() -> Self {
        Self::new()
    }
}
